src/nodes/NodeManager.ts

```diff
@@ -559,183 +771,251 @@ class NodeManager {
+          this.refreshBucketDelayJitter,
+        );
+        await this.taskManager.scheduleTask({
+          handlerId: this.refreshBucketHandlerId,
+          delay: this.refreshBucketDelay + jitter,
+          lazy: true,
+          parameters: [bucketIndex],
+          path: [this.basePath, this.refreshBucketHandlerId, `${bucketIndex}`],
+          priority: 0,
+        });
+      }
     }
+    this.logger.info('Set up refreshBucket tasks');
   }
 
-  private async stopRefreshBucketTimers() {
-    clearTimeout(this.refreshBucketTimer);
-  }
-
-  private refreshBucketTimerReset(timeout: number) {
-    clearTimeout(this.refreshBucketTimer);
-    this.refreshBucketTimer = setTimeout(() => {
-      this.refreshBucketRefreshTimer();
-    }, timeout);
-  }
+  @ready(new nodesErrors.ErrorNodeManagerNotRunning())
+  public async updateRefreshBucketDelay(
+    bucketIndex: number,
+    delay: number = this.refreshBucketDelay,
+    lazy: boolean = true,
+    tran?: DBTransaction,
+  ): Promise<Task> {
+    if (tran == null) {
+      return this.db.withTransactionF((tran) =>
+        this.updateRefreshBucketDelay(bucketIndex, delay, lazy, tran),
+      );
+    }
 
-  public refreshBucketUpdateDeadline(bucketIndex: NodeBucketIndex) {
-    // Update the map deadline
-    this.refreshBucketDeadlineMap.set(
-      bucketIndex,
-      Date.now() + this.refreshBucketTimerDefault,
+    const jitter = nodesUtils.refreshBucketsDelayJitter(
+      delay,
+      this.refreshBucketDelayJitter,
     );
-    // If the bucket was pending a refresh we remove it
-    this.refreshBucketQueueRemove(bucketIndex);
-    if (bucketIndex === this.refreshBucketNext) {
-      // Bucket is same as next bucket, this affects the timer
-      this.refreshBucketRefreshTimer();
+    let foundTask: Task | undefined;
+    let count = 0;
+    for await (const task of this.taskManager.getTasks(
+      'asc',
+      true,
+      [this.basePath, this.refreshBucketHandlerId, `${bucketIndex}`],
+      tran,
+    )) {
+      count += 1;
+      if (count <= 1) {
+        foundTask = task;
+        // Update the first one
+        // total delay is refreshBucketDelay + time since task creation
+        // time since task creation = now - creation time;
+        const delayNew =
+          performance.now() +
+          performance.timeOrigin -
+          task.created.getTime() +
+          delay +
+          jitter;
+        try {
+          await this.taskManager.updateTask(task.id, { delay: delayNew });
+        } catch (e) {
+          if (e instanceof tasksErrors.ErrorTaskMissing) {
+            count -= 1;
+          } else if (!(e instanceof tasksErrors.ErrorTaskRunning)) {
+            throw e;
+          }
+        }
+        this.logger.debug(
+          `Updating refreshBucket task for bucket ${bucketIndex}`,
+        );
+      } else {
+        // These are extra, so we cancel them
+        // TODO: make error
+        task.cancel(Error('TMP, cancel extra tasks'));
```

Comment:
This should be it's own error class.

I would also change the async generator to speicfically deal with the first item, then all subsequent items.

Reply:
I would, but it's not the first task we're interested in, it's the first task that is still running, queued or scheduled. The first one could throw `ErrorTaskMissing` and that wouldn't count as a found task.

I've updated it to use a boolean flag to make it more clear. using a counter wasn't needed.

Reply:
The custom error here can be a special task error that throws into the task handler for cancellation.

Can you create a error here under `src/nodes/errors.ts` for this specific purpose, then document it similar to `ErrorTaskRetry` and `ErrorTaskStop`. See `src/tasks/errors.ts`. These errors are "signals".